Add Pearson4 fit model (original Wikipedia version)

lellid opened the pull request:
#### Description
This adds the Pearson4 fit model (original version from Wikipedia), which is a skewed version of Pearson7.


###### Type of Changes
<!--- What type of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix
- [x] New feature
- [ ] Refactoring / maintenance
- [ ] Documentation / examples


###### Tested on
<!-- Generate version information with this command in the Python shell and copy the output here:
import sys, lmfit, numpy, scipy, asteval, uncertainties
print('Python: {}\n\nlmfit: {}, scipy: {}, numpy: {}, asteval: {}, uncertainties: {}'\
      .format(sys.version, lmfit.__version__, scipy.__version__, numpy.__version__, \
      asteval.__version__, uncertainties.__version__))
-->
Python: 3.10.4 | packaged by conda-forge | (main, Mar 30 2022, 08:38:02) [MSC v.1916 64 bit (AMD64)]
lmfit: 0.0.post2625+g0b9947a.d20220808, scipy: 1.9.0, numpy: 1.23.1, asteval: 0.9.27, uncertainties: 3.1.7


###### Verification <!-- (delete not applicable items) -->
Have you
<!--- Put an `x` in all the boxes that apply OR describe why you think this is unnecessary. -->
- [x] included docstrings that follow PEP 257?
<!-- Please use your favorite linter (e.g., pydocstyle) to check your docstrings. -->
- [x] verified that existing tests pass locally?
<!-- Please run the test-suite locally with pytest and make sure it passes. -->
- [x] verified that the documentation builds locally?
<!-- Please build the documentation (i.e., type make in the "doc" directory) and make sure it finishes. -->
- [x] squashed/minimized your commits and written descriptive commit messages?
<!-- We value a clean history with useful commit messages. Ideally, you will take care of this
         before submitting a PR; otherwise you'll be asked to do so before merging. -->
- [ ] added or updated existing tests to cover the changes? (not neccessary, new model))
- [X] updated the documentation and/or added an entry to the release notes (doc/whatsnew.rst)?
- [ ] added an example? (not neccessary, only a model)

newville commented:
@lellid Thanks!  This looks like a very good start to me, and I'll try to take a more careful look at it.  FWIW, I didn't understand the failures (files were modified by `isort`?).  Maybe @reneeotten can comment on that...

reneeotten commented:
> FWIW, I didn't understand the failures (files were modified by `isort`?). Maybe @reneeotten can comment on that...

I can, the changes/additions to import packages do not follow the guidelines of the pre-commit hooks we use. That's something that could be verified locally before submitting a PR, but in short the diff for `lineshapes.py` in the imports should read like:
```
-from numpy import (arctan, copysign, cos, exp, isclose, isnan, log, pi, real,
-                   sin, sqrt, where)
+from numpy import (arctan, copysign, cos, exp, isclose, isnan, log, log1p, pi,
+                   real, sin, sqrt, where)
+from scipy.special import betaln as betalnfcn
 from scipy.special import erf, erfc
 from scipy.special import gamma as gamfcn
+from scipy.special import loggamma as loggammafcn
 from scipy.special import wofz
```

reneeotten commented on lmfit/lineshapes.py:
```diff
@@ -145,6 +145,24 @@ def moffat(x, amplitude=1, center=0., sigma=1, beta=1.):
+def pearson4(x, amplitude=1.0, center=0.0, sigma=1.0, expon=1.0, skew=0):
+    """Return a Pearson4 lineshape.
+
+    Using the Wikipedia definition:
```
can you please add the link to the Wikipedia where the lineshape is defined (I know we don't have that for the Pearson7 either, and it would probably be a good addition to add that as well). We have the links present for most other lineshapes

newville replied:
I think there may not be a single link to the equation... but a line to the main Pearson distribution page is probably good.

reneeotten commented on lmfit/lineshapes.py:
```diff
@@ -145,6 +145,24 @@ def moffat(x, amplitude=1, center=0., sigma=1, beta=1.):
+    Using the Wikipedia definition:
+
+    pearson4(x, amplitude, center, sigma, expon, skew) =
+        amplitude*|gamma(expon + I skew/2)/gamma(m)|**2/(w*beta(expon-0.5, 0.5)) * (1+arg**2)**(-expon) * exp(-skew * arctan(arg))
```
what is the "I" here (sorry I quickly Googled but didn't find the Wikipedia link immediately nor invested time in looking at the lineshape function)... it should be either a function argument or defined. Also I have a hard time relating the Python function you return to the definition of the Pearson4 lineshape you give, but again as I said I haven't spend much time at it...

reneeotten replied:
I still don't see a clear correlation between the function defined here and what is `return`-ed below as the lineshape function....  is that really only me ;)

reneeotten commented on lmfit/models.py:
```diff
@@ -735,6 +735,60 @@ def guess(self, data, x, negative=False, **kwargs):
+
+    where :math:`\beta` is the beta function (see :scipydoc:`special.beta`).
+    The :meth:`guess` function always gives a starting value of 1.5 for `expon`,
+    and 0 for `skew`. In addition, parameters `fwhmapprox`, `height` and `position` are included as
```
why is the parameter called `fwhmapprox`; this is in other lineshapes sometimes also an approximation - correct? I'd prefer to keep the names consistent and just call it here `fwhm` as well.

reneeotten commented on lmfit/models.py:
```diff
@@ -735,6 +735,60 @@ def guess(self, data, x, negative=False, **kwargs):
+        pars[f'{self.prefix}expon'].set(value=1.5)
+        pars[f'{self.prefix}skew'].set(value=0.0)
+        return update_param_vals(pars, self.prefix, **kwargs)
+
```
as you can see from the Codecov warnings there are no tests added for the new model. Please add a test for it in `test_builtin_models.py`

reneeotten reviewed (changes requested):
thanks for the PR @lellid - a few small comments after a quick look through

newville commented:
@reneeotten Ah, thanks -- "imports out of preferred order"!

reneeotten commented:
> @reneeotten Ah, thanks -- "imports out of preferred order"!

well... the main thing is you're not supposed to have several "import ..... as ....." on one lines as was done in this PR: 
``` from scipy.special import gamma as gamfcn, loggamma as loggammafcn, betaln as betalnfcn```

newville commented on lmfit/models.py:
```diff
@@ -735,6 +735,60 @@ def guess(self, data, x, negative=False, **kwargs):
+               "{prefix:s}expon * log1p(square({prefix:s}skew/(2*{prefix:s}expon))) - {prefix:s}skew * arctan(-{prefix:s}skew/(2*{prefix:s}expon)))")
+        self.set_param_hint('height', expr=fmt.format(tiny, prefix=self.prefix))
+        fmt = ("{prefix:s}center-{prefix:s}sigma*{prefix:s}skew/(2*{prefix:s}expon)")
+        self.set_param_hint('position', expr=fmt.format(prefix=self.prefix))
```
out of curiosity, where do those forms for "fwhm" and "position" come from?  I'm sure we'll get asked at some point.   And what does "position" mean? Is that meant to be "mean position", "position at highest value", or something else?  Those might be nice to have in docstrings.

lellid replied:
Hi,

the `position` is meant to be the position of the highest value (if you put `position` as `x` in pearson4, you would get the `height` value). The formula for position could be easily derived by taking the first derivative and solving it for 0.

I did not find a formula for the `fwhm` value in literature, thus I had to figure out an own one. The approximation is based on the formula for pearson7, but of course with modifications taking the contribution of skew into account. I keep searching for a more accurate one!

newville replied:
@lellid  Thanks.  
For `position`:  OK, maybe the docstring could be a little more explicit about this.   There are a few other lineshapes where `center` might not be the x value of the maximum height.  I agree that in those cases, knowing both are good.   It's just that in my view, you are asking *us* to support (including, answer questions) about this function.   I just want it to be a little clearer because I will definitely forget if it is not documented. ;).

For  `fwhm`:  if approximate is all we have, that's OK.  Maybe describe it as "approximate" in the docstring.

lellid commented:
Hi,

I just updated this PR. I

- added tests (code coverage is not complaining any more)
- put the import statement in lineshapes.py on separate lines
- improved the documentation in models.py
- added Wiki link in lineshapes.py

newville commented:
@lellid @reneeotten if I understand correctly, we're all in favor of merging this, except for some "not-quite-as-beautiful-as-we-like" `import` statements.   Please fix those so we can merge.

newville commented:
@lellid Thanks!  I'll plan to merge in a couple of days to give time for final comments or suggestions.

reneeotten commented on lmfit/models.py:
```diff
@@ -735,6 +735,61 @@ def guess(self, data, x, negative=False, **kwargs):
+    The :meth:`guess` function always gives a starting value of 1.5 for `expon`,
+    and 0 for `skew`. In addition, parameters `fwhm`, `height` and `position` are included as
+    constraints to report approximate full width at half maximum, peak height, and position
+    of the maximal function value, respectively.
```
the last sentence here is pretty much a duplication from the "In addition, ...." above the `math` section - please consolidate this in one place.

reneeotten commented:
> @lellid Thanks! I'll plan to merge in a couple of days to give time for final comments or suggestions.

thanks @lellid - for me the two comments about the docstrings (I left them again above) remain, after making these changes - or explaining why I'm wrong ;) - the commits should be squashed into one - then it's ready-to-go!

newville commented:
@reneeotten @lellid Yes, let's please have consistent code and docstrings!  I think we are happy to merge this, but we kind of need a version we can support, which does include being able to explain the actual definition used. We discussed this earlier at #799,  but this PR was created as a separate PR after that discussion.

To reiterate, I strongly suggest following the Wikipedia definition or having a really clearly documented explanation of why that form was used instead of the Wikipedia definition.

lellid commented:
Hi,

I consolidated the duplication of the docstring.

To the other question about the Wikipedia definition and my implementation:

It is simply a numerical thing:

Of course I can use a naive implementation of $(1+arg^2)^{-m}$, but guess what happens if m is 1000 and arg is 10? Underflow. Or for instance, what is $Gamma(1000)$? Overflow! Finally, both terms together would give some reasonable result, but first, you get underflow multiplied by overflow.

So I logarithmize everything, and put all terms then in the $\exp$ function. For instance,  $(1+arg^2)^{-m} = \exp(-m \times \log(1+arg^2)) =  \exp(-m \times log1p(arg^2))$

I have seen other rather naive implementations in your code (for instance, Pearson7). Maybe, in the future, I will make some pull requests to correct that.

reneeotten commented:
> I consolidated the duplication of the docstring.

thanks!

> To the other question about the Wikipedia definition and my implementation:
> 
> It is simply a numerical thing:
> 
> Of course I can use a naive implementation of (1+arg2)−m, but guess what happens if m is 1000 and arg is 10? Underflow. Or for instance, what is Gamma(1000)? Overflow! Finally, both terms together would give some reasonable result, but first, you get underflow multiplied by overflow.
> 
> So I logarithmize everything, and put all terms then in the exp function. For instance, (1+arg2)−m=exp⁡(−m×log⁡(1+arg2))=exp⁡(−m×log1p(arg2))

That might be clear to you since you've probably stared at that equation for a while and looked at the math, but from a first read it wasn't to me. I also haven't checked that this is all implemented correctly, but I assume that you did... Of course all of this could be considered a "me problem" but perhaps you could just add a note to the function in `lineshapes.py` to mention it; just so that one doesn't need to go back to this PR in a few months/years to remember this ;)

> I have seen other rather naive implementations in your code (for instance, Pearson7). Maybe, in the future, I will make some pull requests to correct that.

Well, naive or not it appears to have been working well for people for quite some time - of course improvements are always welcome.

Thanks again @lellid - this PR should be almost at the finish line.

lellid commented:
Hi,

what I was really wondering about was why you don't have any tests for the function values in your code? At least I haven't seen any. Would make those numerical tweakings a lot more secure, if you can be sure that the values come out that you expect.

newville commented:
@lellid 

> I consolidated the duplication of the docstring.

OK, thanks.  Well, at some point we might fix the capital "I" to "j" (Python's preferred sqrt(-1)).

> To the other question about the Wikipedia definition and my implementation:
>
> It is simply a numerical thing:
>
> Of course I can use a naive implementation of , but guess what happens if m is 1000 and arg is 10? Underflow. Or for instance, what is ? Overflow! Finally, both terms together would give some reasonable result, but first, you get underflow multiplied by overflow.
>
> So I logarithmize everything, and put all terms then in the  function

These functions are kind of prone to under/overflow.  I would not necessarily agree that it is *obvious* that taking `log` and then doing `exp` of that will help in general, but that's to say that it would not help here...

> I have seen other rather naive implementations in your code (for instance, Pearson7). Maybe, in the future, I will make some pull requests to correct that.

Do you mean to use `scipy.special.beta` instead of using the equivalent with `gamma`?   Yeah, that would be okay.

But, I'm not sure that the general approach of "work in log space and then take exponential" is necessarily worth the complication.   Readability definitely counts.

newville commented:
@lellid 

>what I was really wondering about was why you don't have any tests for the function values in your code? 

Probably because we feel like we don't need those specifically.  

The tests of fitting with these functions do test the values of parameters, so if the functions break, that will very likely be reflected in the fit results.  

> At least I haven't seen any. Would make those numerical tweakings a lot more secure, if you can be sure that the values come out that you expect.

Well, I guess we expect that we are not changing the lineshape definitions very much ;)

newville commented:
@reneeotten @lellid I'm just going to squash and merge this.  If there are more issues to resolve, they can be done afterward.
